File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 042888. 오픈채팅방/kotlin/Solution.kt
class Solution {
    fun solution(records: Array<String>): Array<String> {
        val userRepository = HashMap<String, User>()
        val events = ArrayList<Event>()
        var recordSplit: List<String>
        for (record in records) {
            recordSplit = record.split(" ")
            if (recordSplit[0] == "Enter") {
                if (userRepository.containsKey(recordSplit[1])) {
                    userRepository[recordSplit[1]]!!.changeName(recordSplit[2])
                } else {
                    userRepository[recordSplit[1]] = User(recordSplit[1], recordSplit[2])
                }
                events.add(Event(recordSplit[1], true))
            } else if (recordSplit[0] == "Leave") {
                events.add(Event(recordSplit[1], false))
            } else {
                userRepository[recordSplit[1]]!!.changeName(recordSplit[2])
            }
        }
        return events.map {
            if (it.isEnter) {
                "${userRepository[it.userId]!!.name}님이 들어왔습니다."
            } else {
                "${userRepository[it.userId]!!.name}님이 나갔습니다."
            }
        }.toTypedArray()
    }

}

class User (val id: String, var name: String) {

    fun changeName(newName: String) {
        if (newName != name) {
            name = newName;
        }
    }
}

class Event(val userId:String, val isEnter:Boolean)

File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 042888. 오픈채팅방/python/solution.py
def solution(record):
    record = [r.split() for r in record]
    users = {}
    for r in record:
        if len(r) == 2:  # Leave
            continue
        command, uid, nickname = r
        users[uid] = nickname

    result = []
    for r in record:
        if len(r) == 2:
            command, uid = r
        else:
            command, uid, nickname = r

        if command == 'Enter':
            result.append(f'{users[uid]}님이 들어왔습니다.')
        elif command == 'Leave':
            result.append(f'{users[uid]}님이 나갔습니다.')
    return result
